pkg/controllers/disruption/controller.go

```diff
@@ -179,194 +163,59 @@ func (c *Controller) disrupt(ctx context.Context, disruption Method) (bool, erro
 	return true, nil
 }
 
+// executeCommand will do the following, untainting if the step fails.
+// 1. Taint candidate nodes
+// 2. Spin up replacement nodes
+// 3. Add Command to orchestration.Queue to wait to delete the candiates.
 func (c *Controller) executeCommand(ctx context.Context, m Method, cmd Command) error {
-	deprovisioningActionsPerformedCounter.With(map[string]string{
-		actionLabel:        fmt.Sprintf("%s/%s", m.Type(), cmd.Action()),
-		deprovisionerLabel: m.Type(),
-	}).Inc()
 	disruptionActionsPerformedCounter.With(map[string]string{
 		actionLabel:            string(cmd.Action()),
 		methodLabel:            m.Type(),
 		consolidationTypeLabel: m.ConsolidationType(),
 	}).Inc()
 	logging.FromContext(ctx).Infof("disrupting via %s %s", m.Type(), cmd)
 
-	reason := fmt.Sprintf("%s/%s", m.Type(), cmd.Action())
-	if cmd.Action() == ReplaceAction {
-		if err := c.launchReplacementNodeClaims(ctx, m, cmd); err != nil {
-			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
-			// we don't want to disrupt workloads with no way to provision new NodeClaims for them.
-			return fmt.Errorf("launching replacement, %w", err)
-		}
+	stateNodes := lo.Map(cmd.candidates, func(c *Candidate, _ int) *state.StateNode {
+		return c.StateNode
+	})
+	// Cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := state.RequireNoScheduleTaint(ctx, c.kubeClient, true, stateNodes...); err != nil {
+		return multierr.Append(fmt.Errorf("cordoning nodes, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
 	}
+	providerIDs := lo.Map(cmd.candidates, func(c *Candidate, _ int) string { return c.ProviderID() })
+	// We have the new NodeClaims created at the API server so mark the old NodeClaims for deletion
+	c.cluster.MarkForDeletion(providerIDs...)
 
-	for _, candidate := range cmd.candidates {
-		c.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, reason)...)
-
-		if err := nodeclaimutil.Delete(ctx, c.kubeClient, candidate.NodeClaim); err != nil {
-			if !errors.IsNotFound(err) {
-				logging.FromContext(ctx).Errorf("terminating, %s", err)
-			}
-			continue
+	var nodeClaimKeys []nodeclaim.Key
+	var err error
+	if len(cmd.replacements) > 0 {
+		if nodeClaimKeys, err = c.createReplacementNodeClaims(ctx, m, cmd); err != nil {
+			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
+			// we don't want to disrupt workloads with no way to provision new nodes for them.
+			return multierr.Append(fmt.Errorf("launching replacement nodeclaim, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
```

Comment:
This currently isn't "Unmarking" for deletion, but also I think we should move the `MarkForDeletion` statement down so that this statement _shouldn't_ have to

Reply:
Good catch